pkg/kubelet/cm/helpers_linux.go

```diff
@@ -214,20 +211,18 @@ func getCgroupSubsystemsV1() (*CgroupSubsystems, error) {
 
 // getCgroupSubsystemsV2 returns information about the enabled cgroup v2 subsystems
 func getCgroupSubsystemsV2() (*CgroupSubsystems, error) {
-	content, err := ioutil.ReadFile(filepath.Join(util.CgroupRoot, "cgroup.controllers"))
+	allCgroups, err := libcontainercgroups.GetCgroupMounts(false)
 	if err != nil {
-		return nil, err
+		return &CgroupSubsystems{}, err
 	}
-
 	mounts := []libcontainercgroups.Mount{}
-	controllers := strings.Fields(string(content))
-	mountPoints := make(map[string]string, len(controllers))
-	for _, controller := range controllers {
-		mountPoints[controller] = util.CgroupRoot
+	mountPoints := make(map[string]string, len(allCgroups))
+	for _, subsystem := range allCgroups[0].Subsystems {
```

Comment:
Is there guaranteed to always be a '[0]' on `allCgroups`? It seems we shuld check that it's length is `> 1` before accessing it.

Likewise, is `allCgroups[0]` always guaranteed to be the root `cgroup`? It's non-obvious from the code here that this is the case.

It seems we should check that the

Reply:
cgroupv2 has the unified mount point and that never changes.
I think libcontainer guarantees that the function of GetCgroupMounts() returns the unified mount point and the available controllers as allCgroups[0] if it does not return error.
So, I just only checked the error of the function of GetCgroupMounts().

Reply:
I read through the library code and it doesn't appear that there's a case where `allCgroups` would have length < 1 without throwing an error. I agree it'd probably be better to double check this explicitly, but it's probably okay.

Reply:
@kolyshkin ptal

Reply:
In runc `func GetCgroupMounts` has the following comment:

```go
                // TODO: remove cgroupv2 case once all external users are converted
```

as it is semi-meaningless to use it (there's no "mount points" for cgroupv2, there's only one and it is set in stone as /sys/fs/cgroup). So, please don't add more uses for it, especially into cgroupv2-specific code.

Instead you can use `libcontainercgroups.GetAllSubsystems()` which will have the same outcome you're trying to achieve (add "devices" and "freezer" pseudo-controllers).

Reply:
@kolyshkin @mrunalp 
I didn't know the current stuatus of these APIs. I created this commit just about a year ago.
https://github.com/opencontainers/runc/commit/44b75e760ece4fe09a7cd8c56540aff00cad732f

Anyway, I updated to use GetAllSubsystems() instead of GetCgroupMounts().
Thanks!

Reply:
LGTM